File: Python/2015/06.py
from lib import *

input = read_input(2015, 6)


def parse_instructions(instructions):
    out = []

    for line in instructions:
        inst, *coords = re.match(r"^([a-z ]+) (\d+),(\d+) through (\d+),(\d+)$", line).groups()

        out.append((["turn off", "turn on", "toggle"].index(inst), *map(int, coords)))

    return out


lights = [[0 for _ in range(1000)] for _ in range(1000)]
for inst, x1, y1, x2, y2 in parse_instructions(input.splitlines()):
    for y in range(y1, y2 + 1):
        for x in range(x1, x2 + 1):
            lights[y][x] = 1 - lights[y][x] if inst == 2 else inst
print(sum(map(sum, lights)))


lights = [[0 for _ in range(1000)] for _ in range(1000)]
for inst, x1, y1, x2, y2 in parse_instructions(input.splitlines()):
    for y in range(y1, y2 + 1):
        for x in range(x1, x2 + 1):
            lights[y][x] = max(0, lights[y][x] + (inst or -1))
print(sum(map(sum, lights)))

File: Rust/2015/06.rs
#![feature(test)]

use std::sync::LazyLock;

use regex::Regex;

type Input = Vec<Instruction>;

#[derive(Debug)]
struct Instruction {
    cmd: Command,
    start: Point,
    end: Point,
}

#[derive(Debug)]
enum Command {
    On,
    Off,
    Toggle,
}

type Point = (usize, usize);

fn setup(input: &str) -> Input {
    input
        .trim()
        .lines()
        .map(|line| {
            static REGEX: LazyLock<Regex> = LazyLock::new(|| {
                Regex::new(r"^(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)$").unwrap()
            });
            let caps = REGEX.captures(line).unwrap();
            let cmd = match &caps[1] {
                "turn on" => Command::On,
                "turn off" => Command::Off,
                "toggle" => Command::Toggle,
                _ => unreachable!(),
            };

            let start = (caps[2].parse().unwrap(), caps[3].parse().unwrap());
            let end = (caps[4].parse().unwrap(), caps[5].parse().unwrap());

            Instruction { cmd, start, end }
        })
        .collect()
}

fn part1(input: &Input) -> usize {
    let mut grid = vec![false; 1000 * 1000];
    for Instruction { cmd, start, end } in input {
        for x in start.0..=end.0 {
            for y in start.1..=end.1 {
                let i = y * 1000 + x;
                grid[i] = match cmd {
                    Command::On => true,
                    Command::Off => false,
                    Command::Toggle => !grid[i],
                }
            }
        }
    }
    grid.into_iter().filter(|&x| x).count()
}

fn part2(input: &Input) -> u64 {
    let mut grid = vec![0u64; 1000 * 1000];
    for Instruction { cmd, start, end } in input {
        for x in start.0..=end.0 {
            for y in start.1..=end.1 {
                let i = y * 1000 + x;
                grid[i] = match cmd {
                    Command::On => grid[i] + 1,
                    Command::Off => grid[i].saturating_sub(1),
                    Command::Toggle => grid[i] + 2,
                }
            }
        }
    }
    grid.into_iter().sum()
}

aoc::main!(2015, 6);
